sw/host/opentitanlib/src/test_utils/bitbanging.rs

```diff
@@ -0,0 +1,269 @@
+                    .collect(),
+            }
+        }
+
+        fn byte<const SDA: u8, const SCL: u8>(&self, byte: u8, ack: bool) -> Vec<u8> {
+            let mut buffer = (0..8u8)
+                .rev()
+                .flat_map(|bit| {
+                    [
+                        (((byte >> bit) & 0x01) << SDA) | 0x01 << SCL,
+                        (((byte >> bit) & 0x01) << SDA) | 0x00 << SCL,
+                    ]
+                })
+                .collect::<Vec<u8>>();
+            buffer.append(&mut self.ack::<SDA, SCL>(!ack));
+            buffer
+        }
+
+        fn ack<const SDA: u8, const SCL: u8>(&self, nack: bool) -> Vec<u8> {
+            vec![
+                (nack as u8) << SDA | 0x01 << SCL,
+                (nack as u8) << SDA | 0x00 << SCL,
+            ]
+        }
+    }
+    pub struct Encoder<const SDA: u8, const SCL: u8> {}
+    impl<const SDA: u8, const SCL: u8> Encoder<SDA, SCL> {
+        pub fn run(&self, transfer: &[Transfer]) -> Vec<u8> {
+            let mut result: Vec<u8> = transfer
+                .windows(2)
+                .flat_map(|window| {
+                    window[0].parse::<SDA, SCL>(window.get(1) != Some(&Transfer::Stop))
+                })
+                .collect();
+
+            //We missed the last element for using the windows function to peek, so we parse the last element here.
+            result.append(&mut transfer.iter().last().unwrap().parse::<SDA, SCL>(false));
+            result
+        }
+    }
+
+    enum DecodingStage {
+        Start,
+        Bytes,
+    }
+
+    pub struct Decoder<const SDA: u8, const SCL: u8> {
+        pub buffer: [u8; 256],
+    }
+
+    impl<const SDA: u8, const SCL: u8> Decoder<SDA, SCL> {
+        const LOW: u8 = 0;
+        const HIGH: u8 = 1;
+
+        fn sda(sample: u8) -> u8 {
+            (sample >> SDA) & 0x01
+        }
+
+        fn scl(sample: u8) -> u8 {
+            (sample >> SCL) & 0x01
+        }
+
+        fn sample_on_fall_edge<I>(samples: &mut Peekable<I>) -> Result<Option<u8>>
+        where
+            I: Iterator<Item = u8>,
+        {
+            let mut sda = None;
+            loop {
+                if Self::scl(
+                    *samples
+                        .peek()
+                        .ok_or_else(|| anyhow!("Ran out of samples and did not find fall edge"))?,
+                ) == Self::LOW
+                {
+                    return Ok(sda);
+                }
+                sda = Some(
+                    samples
+                        .next()
+                        .ok_or_else(|| anyhow!("Ran out of samples and did not find fall edge"))?,
+                );
+            }
+        }
+
+        fn sample_on_raise_edge<I>(samples: &mut Peekable<I>) -> Option<u8>
+        where
+            I: Iterator<Item = u8>,
+        {
+            loop {
+                let sample = samples.next()?;
+                if Self::scl(sample) == Self::HIGH {
```

Comment:
This is not checking previous SCL is low?

Reply:
It's relying on the caller to make sure that the previous SCL is low.

Reply:
In that case you should document that.

Reply:
Done